huntsman/observatory.py

```diff
@@ -472,43 +355,217 @@ def _create_autoguider(self):
+        # Loop until conditions are met to finish flat-fielding
+        while not all(finished.values()):
+
+            if not safety_func():
+                self.logger.info('Stopping flat-fielding as no longer safe.')
+                break
 
-            flat_obs.field = fields
-            flat_obs.exptime = exptimes
-            flat_obs.min_nexp = len(fields)
-            flat_obs.exp_set_size = len(fields)
+            # Get the FITS headers with a common start time
+            start_time = utils.current_time()
+            fits_headers = self.get_standard_headers(observation=observation)
+            fits_headers['start_time'] = utils.flatten_time(start_time)
+
+            # Take the flat field observations (blocking)
+            current_exptimes = {c: exptimes[c][-1] for c in cameras.keys() if not finished[c]}
+            camera_events = self._take_flat_observation(current_exptimes, observation,
+                                                        fits_headers=fits_headers)
+
+            # Check whether each camera has finished
+            all_too_bright = True
+            all_too_faint = True
+            for cam_name, meta in camera_events.items():
+                current_exptime = current_exptimes[cam_name]
+
+                # Calculate mean counts of last image
+                mean_counts = self._autoflat_mean_counts(meta['filename'], bias)
+                self.logger.debug(f'Mean counts for {cam_name} following'
+                                  f' {current_exptime} exposure: {mean_counts}.')
+
+                # Check if the current exposure is good enough to keep
+                is_too_bright = mean_counts > max_counts
+                is_too_faint = mean_counts < min_counts
+                all_too_bright &= is_too_bright
+                all_too_faint &= is_too_faint
+
+                if is_too_bright or is_too_faint:
+                    self.logger.debug(f'Counts outside valid range for flat field'
+                                      f' image on {cam_name}: {mean_counts}.')
+                    # os.remove(meta['filename']) Need to prevent NGAS push...
+                else:
+                    n_good_exposures[cam_name] += 1
+
+                # Check if we have enough good flats for this camera
+                if n_good_exposures[cam_name] >= max_num_exposures:
+                    self.logger.debug(f'Enough flats taken for {cam_name}.')
+                    finished[cam_name] = True
+                    continue
+
+                # Calculate next exposure time
+                elapsed_time = (utils.current_time() - start_time).sec
+                next_exptime = self._autoflat_next_exptime(
+                        current_exptime, elapsed_time, target_counts, mean_counts)
+                self.logger.debug(f'Suggested exposure time for {cam_name}: {next_exptime}.')
+
+                # Check the next exposure time is within limits
+                if next_exptime > max_exptime:
```

Comment:
Probably doesn't make a huge difference, but still.

```suggestion
                if next_exptime >= max_exptime:
```